### PR TITLE
Simplified build with Azure CLI; various bugfixes

### LINKED ISSUE #107: Hyperparameter a pipeline parameter, not random value
`MLOpsPython/code/training/train.py` picks a hyperparameter value randomly. This is an antipattern in 2 ways: (1) DevOps processes should yield reproducible outputs, (2) hyperparameters should be an AML pipeline parameter.

```
# Randomly pic alpha
alphas = np.arange(0.0, 1.0, 0.05)
alpha = alphas[np.random.choice(alphas.shape[0], 1, replace=False)][0]
```

### LINKED ISSUE #110: Fix scoring endpoint HTTP behavior
1. Scoring endpoint should return json, not stringified json

```
curl -H "Authorization: Bearer $token" $scoringUri \
  -d '{"data":[[1,2,3,4,5,6,7,8,9,10],[10,9,8,7,6,5,4,3,2,1]]}' \
  -H Content-type:application/json
```

Returns:

```
"{\"result\": [5113.099642122809, 3713.6329271385334]}"
```

Expected:
```
{"result": [5113.099642122809, 3713.6329271385334]}
```

Probably we should avoid json.dumps() in:
https://github.com/microsoft/MLOpsPython/blob/41ea9b7765183eca83cf3f193e69777858709d35/code/scoring/score.py#L46

2. Scoring endpoint should not return 200 in case of error

Exceptions are swallowed at
https://github.com/microsoft/MLOpsPython/blob/41ea9b7765183eca83cf3f193e69777858709d35/code/scoring/score.py#L47

The code should not only return the error message but also set an appropriate HTTP return code.

### LINKED ISSUE #113: attach_compute should set exit code after exception
I got a transient service error while retrieving compute. An error was logged but the pipeline continued. The script fails to set an exit code when catching the exception:

https://github.com/microsoft/MLOpsPython/blob/d4a7633152c32335bb43209be3a7d84ff8f4ed0b/ml_service/util/attach_compute.py#L42

```
ComputeTargetException:
	Message: Error occurred retrieving targets:
Response Code: 500
Headers: {'Cache-Control': 'no-cache', 'Pragma': 'no-cache', 'Content-Length': '294', 'Content-Type': 'application/json; charset=utf-8', 'Expires': '-1', 'x-ms-ratelimit-remaining-subscription-reads': '11999', 'x-ms-failure-cause': 'service', 'Request-Context': 'appId=cid-v1:6a27ce65-5555-41a3-85f7-b7a1ce31fd6b', 'x-ms-request-id': '|eaf2c3b2dc39574abbc4581269bde67b.8a46c099_1.1bf9b6d7_', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'x-ms-correlation-request-id': 'f3c9e09d-9986-43dc-a81e-9328fbb63de9', 'x-ms-routing-request-id': 'WESTEUROPE:20191126T054559Z:f3c9e09d-9986-43dc-a81e-9328fbb63de9', 'X-Content-Type-Options': 'nosniff', 'Date': 'Tue, 26 Nov 2019 05:45:59 GMT', 'Connection': 'close'}
Content: b'{"error":{"code":"InternalServerError","message":"An internal server error occurred. Please try again. If the problem persists, contact support","innererror":{"clientRequestId":"888141ac-de9b-418a-87e7-987fa829b2ae","serviceRequestId":"|eaf2c3b2dc39574abbc4581269bde67b.8a46c099_1.1bf9b6d7_"}}}'
	InnerException None
	ErrorResponse 
{
    "error": {
        "message": "Error occurred retrieving targets:\nResponse Code: 500\nHeaders: {'Cache-Control': 'no-cache', 'Pragma': 'no-cache', 'Content-Length': '294', 'Content-Type': 'application/json; charset=utf-8', 'Expires': '-1', 'x-ms-ratelimit-remaining-subscription-reads': '11999', 'x-ms-failure-cause': 'service', 'Request-Context': 'appId=cid-v1:6a27ce65-5555-41a3-85f7-b7a1ce31fd6b', 'x-ms-request-id': '|eaf2c3b2dc39574abbc4581269bde67b.8a46c099_1.1bf9b6d7_', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'x-ms-correlation-request-id': 'f3c9e09d-9986-43dc-a81e-9328fbb63de9', 'x-ms-routing-request-id': 'WESTEUROPE:20191126T054559Z:f3c9e09d-9986-43dc-a81e-9328fbb63de9', 'X-Content-Type-Options': 'nosniff', 'Date': 'Tue, 26 Nov 2019 05:45:59 GMT', 'Connection': 'close'}\nContent: b'{\"error\":{\"code\":\"InternalServerError\",\"message\":\"An internal server error occurred. Please try again. If the problem persists, contact support\",\"innererror\":{\"clientRequestId\":\"888141ac-de9b-418a-87e7-987fa829b2ae\",\"serviceRequestId\":\"|eaf2c3b2dc39574abbc4581269bde67b.8a46c099_1.1bf9b6d7_\"}}}'"
    }
}
An error occurred trying to provision compute.
```

### LINKED ISSUE #117: Deployment to QA failed
ENOENT: no such file or directory, stat '/home/vsts/work/r1/a/_ci-build/mlops-pipelines/code/scoring/deployment_config_aci.yml'

******************************************************************************
2019-11-26T12:52:16.5473405Z ##[section]Starting: Azure ML Model Deploy
2019-11-26T12:52:16.5478581Z ==============================================================================
2019-11-26T12:52:16.5478685Z Task         : AzureML Model Deploy
2019-11-26T12:52:16.5478763Z Description  : AzureML Deploy Model
2019-11-26T12:52:16.5478820Z Version      : 0.8.19
2019-11-26T12:52:16.5478891Z Author       : ms-air-aiagility
2019-11-26T12:52:16.5478965Z Help         : https://marketplace.visualstudio.com/items?itemName=ms-air-aiagility.private-vss-services-azureml&ssr=false#overview
2019-11-26T12:52:16.5479065Z ==============================================================================
2019-11-26T12:52:16.8866440Z { Error: ENOENT: no such file or directory, stat '/home/vsts/work/r1/a/_ci-build/mlops-pipelines/code/scoring/deployment_config_aci.yml'
2019-11-26T12:52:16.8866750Z     at Error (native)
2019-11-26T12:52:16.8866819Z     at Object.fs.statSync (fs.js:1000:11)
2019-11-26T12:52:16.8869025Z     at AMLDeployParams.validateFileAndExtension (/home/vsts/work/_tasks/AMLModelDeploy_d9237456-3431-4c1d-b7bf-d6b937425f0f/0.8.19/operations/AMLDeployParams.js:214:17)
2019-11-26T12:52:16.8869853Z     at AMLDeployParams.validateDeployConfig (/home/vsts/work/_tasks/AMLModelDeploy_d9237456-3431-4c1d-b7bf-d6b937425f0f/0.8.19/operations/AMLDeployParams.js:162:14)
2019-11-26T12:52:16.8870255Z     at AMLDeployParams.<anonymous> (/home/vsts/work/_tasks/AMLModelDeploy_d9237456-3431-4c1d-b7bf-d6b937425f0f/0.8.19/operations/AMLDeployParams.js:86:30)
2019-11-26T12:52:16.8870668Z     at step (/home/vsts/work/_tasks/AMLModelDeploy_d9237456-3431-4c1d-b7bf-d6b937425f0f/0.8.19/operations/AMLDeployParams.js:33:23)
2019-11-26T12:52:16.8871044Z     at Object.next (/home/vsts/work/_tasks/AMLModelDeploy_d9237456-3431-4c1d-b7bf-d6b937425f0f/0.8.19/operations/AMLDeployParams.js:14:53)
2019-11-26T12:52:16.8871441Z     at /home/vsts/work/_tasks/AMLModelDeploy_d9237456-3431-4c1d-b7bf-d6b937425f0f/0.8.19/operations/AMLDeployParams.js:8:71
2019-11-26T12:52:16.8871823Z     at __awaiter (/home/vsts/work/_tasks/AMLModelDeploy_d9237456-3431-4c1d-b7bf-d6b937425f0f/0.8.19/operations/AMLDeployParams.js:4:12)
2019-11-26T12:52:16.8872275Z     at AMLDeployParams.getDeploymentParameters (/home/vsts/work/_tasks/AMLModelDeploy_d9237456-3431-4c1d-b7bf-d6b937425f0f/0.8.19/operations/AMLDeployParams.js:81:16)
2019-11-26T12:52:16.8872510Z   errno: -2,
2019-11-26T12:52:16.8872752Z   code: 'ENOENT',
2019-11-26T12:52:16.8872959Z   syscall: 'stat',
2019-11-26T12:52:16.8873249Z   path: '/home/vsts/work/r1/a/_ci-build/mlops-pipelines/code/scoring/deployment_config_aci.yml' }
2019-11-26T12:52:16.8948140Z ##[error]ENOENT: no such file or directory, stat '/home/vsts/work/r1/a/_ci-build/mlops-pipelines/code/scoring/deployment_config_aci.yml'
2019-11-26T12:52:16.8961783Z Reverting to original directory.....
2019-11-26T12:52:16.8961920Z (node:3335) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): Error: Failed popd: popd: directory stack empty
2019-11-26T12:52:16.8961998Z 
2019-11-26T12:52:16.8962110Z popd: directory stack empty
2019-11-26T12:52:16.8962149Z 
2019-11-26T12:52:16.8964165Z ##[section]Finishing: Azure ML Model Deploy

### LINKED ISSUE #118: PipelineParameters format in call to Azure ML Extension
https://github.com/microsoft/MLOpsPython/blob/d36019858e06d33b55ea60312ffa8a3edfb945ba/.pipelines/azdo-ci-build-train.yml#L71

The correct format to have this picked up by the pipeline (confirmed with the extension development team) is:
```
PipelineParameters: '"ParameterAssignments": {"model_name": "sklearn_regression_model.pkl"}'
```

which is results in the following call:

````
POST https://westeurope.aether.ms/api/v1.0/subscriptions/.../workspaces/aml-myws/PipelineRuns/PipelineSubmit/75a219db-0ec0-4f93-a6ac-4fc402da809f
Request body: {
  ...
  "ExperimentName": "mlopspython",
  "RunSource": "PipelineRun",
  "ParameterAssignments": {
    "model_name": "sklearn_regression_model.pkl"
  }
}
```

### LINKED ISSUE #119: Model build fails for Azure Web App for containers
Following the instructions to "Deploy the trained model to Azure Web App for containers", the build runs on a standard agent, so the SDK required by the build script is missing:

```
2019-11-28T07:59:58.9581465Z     from azureml.core import Workspace
2019-11-28T07:59:58.9581884Z ImportError: No module named azureml.core
```

### PR DESCRIPTION
**Pull request by algattik**
* Use Azure CLI tasks to remove need for variables SP_APP_ID, SP_APP_SECRET, SUBSCRIPTION_ID, TENANT_ID
* Updated getting_started.md to point out variable path to artifact (closes #117)
* Made random hyperparameter (ridge regression alpha) a pipeline parameter, setting its value in the DevOps pipeline (closes #107)
* Changed unit test to test actual training code (partially solves #74)
* Fixed: attach_compute should set exit code after exception (closes #113)
* Fixed scoring endpoint HTTP behavior (closes #110)
* Fixed PipelineParameters format in call to Azure ML Extension (closes #118)
* Fixed Model build environment for Azure Web App for containers (closes #119)

### CODE REVIEW

File: .pipelines/azdo-base-pipeline.yml [hunk 1]
```diff
@@ -10,6 +10,7 @@ steps:
   enabled: 'true'
   
 - script: |
+   pip install --user -r $(Build.SourcesDirectory)/tests/requirements.txt
```
Comment (dtzar) [inline .pipelines/azdo-base-pipeline.yml]: Let's put these into the main requirements.txt file

File: .pipelines/azdo-ci-build-train.yml [hunk 2]
```diff
@@ -77,22 +91,28 @@ stages:
...
-        SP_APP_SECRET: '$(SP_APP_SECRET)'
+    - task: AzureCLI@1
+      inputs:
+        azureSubscription: '$(WORKSPACE_SVC_CONNECTION)'
+        scriptLocation: inlineScript
+        inlineScript: |
+          set -e # fail on error
+          export SUBSCRIPTION_ID=$(az account show --query id -o tsv)
+          python $(Build.SourcesDirectory)/ml_service/pipelines/verify_train_pipeline.py --build_id $(Build.BuildId) --model_name "$(MODEL_NAME)"
+          echo "##vso[task.setvariable variable=MODEL_VERSION;isOutput=true]$(cat model_version.txt)"
```
Comment (dtzar) [inline .pipelines/azdo-ci-build-train.yml]: Please fix this output on this step.  You can [see it in this build](https://aidemos.visualstudio.com/MLOps/_build/results?buildId=3190&view=logs&j=bdf7ed77-c6a0-5b06-edd2-dd86f84ec0e0&t=02aa005f-1b50-573a-e7fc-2dcee0005b34).

```
Model was registered for this build.
cat: model_version.txt: No such file or directory
```

File: docs/getting_started.md [hunk 3]
```diff
@@ -300,19 +284,33 @@ config. To learn more on how to create a container with AML SDK click
 [here](https://docs.microsoft.com/en-us/python/api/azureml-core/azureml.core.image.image.image?view=azure-ml-py#create-workspace--name--models--image-config-).
 
 Below is release pipeline with two tasks one to create an image using the above
-script and second is the deploy the image to Web App for containers
-![release_webapp](./images/release-webapp-pipeline.PNG).
+script and second is the deploy the image to Web App for containers.
+
+![release_webapp](./images/release-webapp-pipeline.PNG)
+
+In the Variables tab, link the pipeline to your variable group (`devopsforai-aml-vg`). In the variable group definition, add the following variables:
```
Comment (dtzar) [inline docs/getting_started.md]: These should already be picked up by the variables template file.  If you want to add instructions for people to modify that file, then no problem else remove this section.

File: ml_service/pipelines/run_train_pipeline.py [hunk 4]
```diff
@@ -1,24 +1,34 @@
...
 from env_variables import Env
 
 
 def main():
+
+    parser = argparse.ArgumentParser("register")
+    parser.add_argument(
+        "--output_pipeline_id_file",
+        type=str,
+        help="Name of a file to write pipeline ID to"
```
Comment (dtzar) [inline ml_service/pipelines/run_train_pipeline.py]: Would be nice to set a default file name so no need to pass a filename with the arg.

File: ml_service/util/create_scoring_image.sh [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
+
+set -euo pipefail # strict mode, fail on error
+set -x # verbose
+
+docker run \
```
Comment (dtzar) [inline ml_service/util/create_scoring_image.sh]: Eugene already has a pipeline file to do this and not the best idea to do docker-in-docker.  I'd prefer you remove this file and we can work into a different PR.

File: tests/requirements.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+scikit-learn==0.21.3
```
Comment (dtzar) [inline tests/requirements.txt]: Again, please remove/merge to main requirements.txt file.

File: ml_service/pipelines/verify_train_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    parser = argparse.ArgumentParser("register")
+    parser.add_argument(
+        "--build_id",
+        type=str,
+        help="The Build ID of the build triggering this pipeline run",
+    )
+    parser.add_argument(
+        "--model_name",
+        type=str,
+        help="Name of the Model"
```
Comment (dtzar) [inline ml_service/pipelines/verify_train_pipeline.py]: Let's put a default value here if none specified.  Also default value in the variables template.

File: .pipelines/azdo-ci-build-train.yml (hunk 2 above)
Comment (dtzar) [reply .pipelines/azdo-ci-build-train.yml]: This is likely due to me not specifying a MODEL_VERSION variable.  If you set the default value of the variable to be `model_version.txt` then we'll probably be good.

File: docs/getting_started.md (hunk 3 above)
Comment (algattik) [reply docs/getting_started.md]: This is the doc for the release pipeline, which if I'm not mistaken the user must create by hand from scratch. (This is "resolved" in PR #122 that replace the release pipeline with a multi-stage build pipeline)

File: ml_service/pipelines/run_train_pipeline.py (hunk 4 above)
Comment (algattik) [reply ml_service/pipelines/run_train_pipeline.py]: Adding it as default, but still I prefer to make its value explicit in the call below, as I think it makes the logic of the pipeline clearer. Convention over configuration is good up to the point where it makes threads hard to follow :)

```
          python $(Build.SourcesDirectory)/ml_service/pipelines/run_train_pipeline.py --output_pipeline_id_file "pipeline_id.txt" --skip_train_execution
          # Set AMLPIPELINEID variable for next AML Pipeline task in next job
          AMLPIPELINEID="$(cat pipeline_id.txt)"
```

File: ml_service/util/create_scoring_image.sh (hunk 5 above)
Comment (algattik) [reply ml_service/util/create_scoring_image.sh]: This is run as part of the *release pipeline* directly on the agent, so no docker-in-docker.

I agree this script is pretty ugly-looking. It is removed anyway with PR #122.

If I just remove the file, the getting_started file will not have complete instructions to release a webapp...

File: .pipelines/azdo-base-pipeline.yml (hunk 1 above)
Comment (algattik) [reply .pipelines/azdo-base-pipeline.yml]: Done

File: .pipelines/azdo-ci-build-train.yml (hunk 2 above)
Comment (algattik) [reply .pipelines/azdo-ci-build-train.yml]: Line 102 was actually dead code (belongs in PR #122), I've removed it.

Also updated line 59 (https://github.com/algattik/MLOpsPython/blob/0cf896392717292a66fd8ca02164adecc605e4f1/.pipelines/azdo-ci-build-train.yml#L59) where a similar logic is used, so that the failure of the `cat` command fails the whole job.

File: tests/requirements.txt (hunk 6 above)
Comment (algattik) [reply tests/requirements.txt]: Done. Note that running the build from the PR now requires the image at mcr.microsoft.com/mlops/python:latest to have been updated accordingy.

File: ml_service/pipelines/verify_train_pipeline.py (hunk 7 above)
Comment (algattik) [reply ml_service/pipelines/verify_train_pipeline.py]: Great point. I've removed the argument and the script just takes the value from e.model_name.

**Comment by algattik** [comment]
Comments addressed with code updates or answers above. 

Because of changes to requirements.txt, running the build from the PR now requires the image at mcr.microsoft.com/mlops/python:latest to have been updated accordingly. This is why the unit tests are now failing on the PR check. I've verified a modified version of the PR where the pipeline runs on a container deployed in a private registry, built from environment_setup/Dockerfile.
